### PR TITLE
Replication Ruby Enumerable functions

### PR DESCRIPTION
**Pull request by kunmi02**
Replicating the following Ruby Enumerable functions:

my_each
my_each_with_index
my_select
my_all
my_none
my_count
my_map
my_inject
multiply_els
Tests for each method and linter files have been included.

Thank You.

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+puts([1, 2, 3].my_map(proc1))
+
+puts('## Both Proc and Block')
+puts([1, 2, 3].my_map(proc1) { |item| item + 1 })
+
+puts('## Non Proc and Block')
+puts([1, 2, 3].my_map('Hello') { |item| item + 1 })
+
+puts('## Only Block')
+puts([1, 2, 3].my_map { |item| item + 1 })
```
Comment (PhillipUg) [inline main.rb]: - [ ] These test cases should not be committed with the code. Please remove them

**Review by PhillipUg — CHANGES_REQUESTED** [review]
Hey **Azeez Azeez & Suyash Fowdar**, 👋

You have done a great job on this project so far! Your test coverage was good.

However, there a few things that still need to be worked on before we can approve this request 👇🏿 

### Changed Required ♻️ 

- [ ] You have some rubocop errors. Please fix these before making your next submission.

- [ ] You have submitted your code together with some test cases at the bottom of your `main.rb` file. Please remove these test cases and submit just the code before making your next submission. (See inline comment)

- [ ] Your test cases are not exhaustive enough, below, I am going to provide you with some custom test cases. Use these to improve your enumerable methods.

```
puts multiply_els([2, 4, 5]) # => 40
puts

# 1. my_each
puts 'my_each'
puts '-------'
puts [1, 2, 3].my_each { |elem| print "#{elem + 1} " } # => 2 3 4
puts

# 2. my_each_with_index
puts 'my_each_with_index'
puts '------------------'
print [1, 2, 3].my_each_with_index { |elem, idx| puts "#{elem} : #{idx}" } # => 1 : 0, 2 : 1, 3 : 2
puts

# 3. my_select
puts 'my_select'
puts '---------'
p [1, 2, 3, 8].my_select(&:even?) # => [2, 8]
p [0, 2018, 1994, -7].my_select { |n| n > 0 } # => [2018, 1994]
p [6, 11, 13].my_select(&:odd?) # => [11, 13]
puts

# 4. my_all? (example test cases)
puts 'my_all?'
puts '-------'
p [3, 5, 7, 11].my_all?(&:odd?) # => true
p [-8, -9, -6].my_all? { |n| n < 0 } # => true
p [3, 5, 8, 11].my_all?(&:odd?) # => false
p [-8, -9, -6, 0].my_all? { |n| n < 0 } # => false
# test cases required by tse reviewer
p [1, 2, 3, 4, 5].my_all? # => true
p [1, 2, 3].my_all?(Integer) # => true
p %w[dog door rod blade].my_all?(/d/) # => true
p [1, 1, 1].my_all?(1) # => true
puts

# 5. my_any? (example test cases)
puts 'my_any?'
puts '-------'
p [7, 10, 3, 5].my_any?(&:even?) # => true
p [7, 10, 4, 5].my_any?(&:even?) # => true
p %w[q r s i].my_any? { |char| 'aeiou'.include?(char) } # => true
p [7, 11, 3, 5].my_any?(&:even?) # => false
p %w[q r s t].my_any? { |char| 'aeiou'.include?(char) } # => false
# test cases required by tse reviewer
p [3, 5, 4, 11].my_any? # => true
p [1, nil, false].my_any?(1) # => true
p [1, nil, false].my_any?(Integer) # => true
p %w[dog door rod blade].my_any?(/z/) # => false
p [1, 2, 3].my_any?(1) # => true
puts

# 6. my_none? (example test cases)
puts 'my_none?'
puts '--------'
p [3, 5, 7, 11].my_none?(&:even?) # => true
p %w[sushi pizza burrito].my_none? { |word| word[0] == 'a' } # => true
p [3, 5, 4, 7, 11].my_none?(&:even?) # => false
p %w[asparagus sushi pizza apple burrito].my_none? { |word| word[0] == 'a' } # => false
# test cases required by tse reviewer
p [1, 2, 3].my_none? # => false
p [1, 2, 3].my_none?(String) # => true
p [1, 2, 3, 4, 5].my_none?(2) # => false
p [1, 2, 3].my_none?(4) # => true
puts

# 7. my_count (example test cases)
puts 'my_count'
puts '--------'
p [1, 4, 3, 8].my_count(&:even?) # => 2
p %w[DANIEL JIA KRITI dave].my_count { |s| s == s.upcase } # => 3
p %w[daniel jia kriti dave].my_count { |s| s == s.upcase } # => 0
# test cases required by tse reviewer
p [1, 2, 3].my_count # => 3
p [1, 1, 1, 2, 3].my_count(1) # => 3
puts

# 8. my_map
puts 'my_map'
puts '------'
p [1, 2, 3].my_map { |n| 2 * n } # => [2,4,6]
p %w[Hey Jude].my_map { |word| word + '?' } # => ["Hey?", "Jude?"]
p [false, true].my_map(&:!) # => [true, false]
my_proc = proc { |num| num > 10 }
p [18, 22, 5, 6].my_map(my_proc) { |num| num < 10 } # => true true false false
puts

# 9. my_inject
puts 'my_inject'
puts '---------'
p [1, 2, 3, 4].my_inject(10) { |accum, elem| accum + elem } # => 20
p [1, 2, 3, 4].my_inject { |accum, elem| accum + elem } # => 10
p [5, 1, 2].my_inject('+') # => 8
p (5..10).my_inject(2, :*) # should return 302400
p (5..10).my_inject(4) { |prod, n| prod * n } # should return 604800
```
#### Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack as **Phillip Musiime**

I would be happy to help! 👩‍💻 

Happy coding 🤓

**Review by ebeagusamuel — CHANGES_REQUESTED** [review]
# Changes Required :recycle: 
Hello team :wave: 
I can see you have implemented some of the changes the previous TSE requested. Good work on that.
However, there are still some improvements to be made. Your test cases are currently not as exhaustive as it should be. 
The whole idea of this particular project is to totally replicate the behavior of the ruby enumerable methods you were asked to recreate. What that means is that your custom enumerable method should behave exactly like that of the original ruby method, such that when passed the same arguments, they both return exactly the same result.
I know this project can be exhausting and challenging, but I think it will be of great help and insight to you if you take the time to check the ruby docs and study how those methods work. That way you will be better informed on the best ways to go about recreating the behavior of those methods.
Also, the previous TSE left you some test cases you can use for testing your project. In addition to that, I would suggest you also use the original method to test your custom methods. for e.g, if you want to test say your my_select method, you should also pass the same arguments to the original select method and see if you get the same result.

Please make another review request when you are done and happy coding
slack(@ebeagusamuel)

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,118 @@
+# rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/ModuleLength
+module Enumerable
+  def my_each
+    i = 0
+    while i < to_a.length
+      yield(to_a[i])
+      i += 1
+    end
+  end
```
Comment (PhillipUg) [inline main.rb]: - [x] returns Enumerator if no block is given

- [x] when self is an array or a range it returns the array or range itself after calling the given block once for each element in self. You can solve this by returning self at the end.

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+  def my_each_with_index
+    i = 0
+    my_each do |item|
+      yield(item, i)
+      i += 1
+    end
+    self
+  end
```
Comment (PhillipUg) [inline main.rb]: - [x] returns an enumerator if no block is given

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+  def my_select
+    result = []
+    each do |e|
+      result << e if yield(e)
+    end
+    result
+  end
```
Comment (PhillipUg) [inline main.rb]: - [ ] returns an enumerator if no block is given

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+      my_each { |n| return false if n.nil? || n == false }
+    elsif !data.nil? && (data.is_a? Class)
+      my_each { |n| return false if n.class != data }
+    elsif !data.nil? && (data.is_a? Regexp)
+      my_each { |n| return false unless data.match(n) }
+    else
+      my_each { |n| return false if n != data }
+    end
+    true
+  end
```
Comment (PhillipUg) [inline main.rb]: - [x] when a class is passed as an argument, returns true if all of the elements in the collection are members of the `Numeric` class

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+      my_each { |item| return true if yield(item) }
+    elsif data.nil?
+      my_each { |n| return true unless n.nil? }
+    elsif data.is_a? Class
+      my_each { |n| return true if n.instance_of?(data) }
+    else
+      my_each { |n| return true if n == data }
+    end
+    false
+  end
```
Comment (PhillipUg) [inline main.rb]: - [x] when no block or argument is given, returns false if at least one of the elements in the collection is not true

- [ ] when a class is passed as an argument, returns true if all of the elements in the collection are members of the Numeric class

- [x] when a Regex is passed as an argument, returns true if any of the collection matches the Regex

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    count = 0
+    if block_given?
+      my_each { |item| count += 1 if yield(item) }
+    elsif !data.nil?
+      my_each { |item| count += 1 if item == data }
+    else
+      count = length
+    end
+    count
+  end
```
Comment (PhillipUg) [inline main.rb]: - [ ] returns the number of items in a range

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    new_array = []
+    if !proc1.nil? && proc1.respond_to?(:call)
+      my_each { |item| new_array.push(proc1.call(item)) }
+    elsif block_given?
+      my_each { |item| new_array.push(yield(item)) }
+    else
+      new_array = self
+    end
+    new_array
+  end
```
Comment (PhillipUg) [inline main.rb]: - [ ] returns an Enumerator if no block is given

File: main.rb [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+      if data[1]
+        net = data[0]
+        my_each { |item| net = net.send(data[1], item) }
+      else
+        net = first
+        drop(1).my_each { |item| net = net.send(data[0], item) }
+      end
+    end
+    net
+  end
```
Comment (PhillipUg) [inline main.rb]: - [ ] raises a "LocalJumpError" when no block or argument is given

**Review by PhillipUg — CHANGES_REQUESTED** [review]
Hey team, 👋

You have made great progress on this project sofar!

However, there are still some changes 👇 you need to implement to approve this request 🔍
## Changes Required ♻️ 

I have left you some inline comments, Please check them out below 👇🏿 👇🏿 

#### Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack as **Phillip Musiime**

I would be happy to help! 👩‍💻 

Happy coding 🤓

**Review by layzrd — CHANGES_REQUESTED** [review]
### Status: Require changes  ♲
Hi, @kunmi02 and @@Krishnzzz, I am [Abdulaziz Redi Ali](https://azizali.ml/), your TSE in this PR.  👏 👏  ✨ I applaud you for your hard work in implementing custom enumerable methods. Most requirements have done 👌, Kindly fix the following issues.

[REQUIRED]

- [ ] #my_each method should return the array/range itself after calling the given block once for each element in self

	e.g. 
	```
	ary = [1,2,3]
	new_ary = array.each(&block)
	# new_ary == ary
	```
	```
	range = (1..3)
	new_range = range.each(&block)
	# new_range == range
	```
 

- [ ] #my_all? and #my_any?  should return true when a class is passed as an argument, and all of the collection is a member of a given class 

	E.g. 
	```
	[2 ,3, 4].my_all?(Numeric)  # => true
	[1, 2, 3].my_all?(Integer) # => true
	```
  

- [ ] #my_any? should return false when no block or argument is given, and at least one of the collection is not true

- [ ] #my_count should return the number of items in enum through enumeration::range

      E.g. `(1..5).my_count`  should return 5

- [ ] #my_map should return an Enumerator if no block is given

[OPTIONAL]

- [ ] Enable all rubocop rules

- [ ] Refactor your complex methods(my_all?, my_none?, my_all?) by extracting a code fragment into a new method

Once you fix the issues, kindly submit another code review. ♲

Happy coding 👨‍💻
[Abdulaziz Redi Ali](https://azizali.ml/) | [Contact me on Slack](https://microverse-students.slack.com/team/U011SD6JUH2)

**Comment by kunmi02** [comment]
HI @abredi this test `[2 ,3, 4].my_all?(Numeric)  # => true`  will fail. Please check the https://docs.ruby-lang.org/en/2.4.0/Numeric.html#:~:text=Numeric%20is%20the%20class%20from,is%20always%20passed%20by%20value

**Review by Temmarie — APPROVED** [review]
## Status: Approved :heavy_check_mark:

### Hi Team!

### Thank you for making all of the required changes, your project is approved :clap: :clap: :clap: :tada:

### You may merge, good luck with your next projects! :rocket:
### Happy Coding! :+1:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

**Comment by layzrd** [comment]
> HI @abredi this test `[2 ,3, 4].my_all?(Numeric) # => true` will fail. Please check the https://docs.ruby-lang.org/en/2.4.0/Numeric.html#:~:text=Numeric%20is%20the%20class%20from,is%20always%20passed%20by%20value

Hello @kunmi02, sorry for the late response.  I hope it will improve the project/skills.

`
[2 ,3, 4].all?(Numeric)
`
is returning `true` so, your custom implementation `[2 ,3, 4].my_all?(Numeric)` should return the same. That's why I mentioned that. 
For more detail check the official documentation for ruby enumerable `all?` https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-all-3F

[Abdulaziz Redi Ali](https://azizali.ml/) | [Contact me on Slack](https://microverse-students.slack.com/team/U011SD6JUH2)

**Pull request merged**